File: mci/layout.py
from __future__ import annotations

from typing import List, Optional

from .types import CardRect, ColumnLayout, CurriculumFile, LayoutData

# ─────────────────────────────────────────────────────────────────────────────
# Constantes de dimensionamento
# ─────────────────────────────────────────────────────────────────────────────

CARD_WIDTH = 140
CARD_HEIGHT = 60
COL_GAP = 60  # espaço horizontal entre colunas (canal para setas)
ROW_GAP = 24  # espaço vertical entre cartões (canal para setas)
COL_PADDING = 10  # margem interna horizontal da coluna
PAGE_MARGIN = 20  # margem externa da grade
HEADER_H = 80  # altura do cabeçalho do curso
COL_HEADER_H = 48  # altura do cabeçalho de cada coluna (nível)
COL_FOOTER_H = 32  # altura do rodapé de cada coluna (total créditos)

# ─────────────────────────────────────────────────────────────────────────────
# Layout
# ─────────────────────────────────────────────────────────────────────────────


def compute_layout(data: CurriculumFile) -> LayoutData:
    columns: List[ColumnLayout] = []

    for level in range(1, data.curriculum.levels + 1):
        level_courses = [c for c in data.courses if c.level == level]
        total_credits = sum(c.credits for c in level_courses)
        col_x = (level - 1) * (CARD_WIDTH + COL_GAP)
        cards_start_y = COL_HEADER_H + ROW_GAP

        cards = [
            CardRect(
                course_code=course.code,
                x=col_x,
                y=cards_start_y + i * (CARD_HEIGHT + ROW_GAP),
                width=CARD_WIDTH,
                height=CARD_HEIGHT,
            )
            for i, course in enumerate(level_courses)
        ]

        columns.append(
            ColumnLayout(level=level, x=col_x, total_credits=total_credits, cards=cards)
        )

    canvas_width = (
        data.curriculum.levels * CARD_WIDTH + (data.curriculum.levels - 1) * COL_GAP
    )

    max_cards = max((len(col.cards) for col in columns), default=0)
    canvas_height = (
        COL_HEADER_H
        + ROW_GAP
        + max_cards * (CARD_HEIGHT + ROW_GAP)
        + COL_FOOTER_H
        + PAGE_MARGIN
    )

    return LayoutData(
        canvas_width=canvas_width, canvas_height=canvas_height, columns=columns
    )


def find_card(layout: LayoutData, course_code: str) -> Optional[CardRect]:
    for col in layout.columns:
        for card in col.cards:
            if card.course_code == course_code:
                return card
    return None

File: src/layout.ts
import { CurriculumFile } from './types';
import { CardRect, ColumnLayout, LayoutData } from './types';

// ─── Constantes de dimensionamento ───────────────────────────────────────────

export const CARD_WIDTH   = 140;
export const CARD_HEIGHT  = 84;
export const COL_GAP      = 60;   // espaço horizontal entre colunas (canal para setas)
export const ROW_GAP      = 24;   // espaço vertical entre cartões (canal para setas)
export const COL_PADDING  = 10;   // margem interna horizontal da coluna
export const PAGE_MARGIN  = 20;   // margem externa da grade
export const HEADER_H     = 80;   // altura do cabeçalho do curso
export const COL_HEADER_H = 48;   // altura do cabeçalho de cada coluna (nível)
export const COL_FOOTER_H = 32;   // altura do rodapé de cada coluna (total créditos)

// ─── Ponto de entrada ─────────────────────────────────────────────────────────

export function computeLayout(data: CurriculumFile): LayoutData {
  const { curriculum, courses } = data;

  const columns: ColumnLayout[] = [];

  for (let level = 1; level <= curriculum.levels; level++) {
    const levelCourses = courses
      .filter(c => c.level === level);

    const totalCredits = levelCourses.reduce((sum, c) => sum + c.credits, 0);

    const colX = (level - 1) * (CARD_WIDTH + COL_GAP);
    const cardsStartY = COL_HEADER_H + ROW_GAP;

    const cards: CardRect[] = levelCourses.map((course, i) => ({
      courseCode: course.code,
      x: colX,
      y: cardsStartY + i * (CARD_HEIGHT + ROW_GAP),
      width: CARD_WIDTH,
      height: CARD_HEIGHT,
    }));

    columns.push({ level, x: colX, totalCredits, cards });
  }

  const canvasWidth =
    curriculum.levels * CARD_WIDTH +
    (curriculum.levels - 1) * COL_GAP;

  const maxCards = Math.max(...columns.map(col => col.cards.length));
  const canvasHeight =
    COL_HEADER_H +
    ROW_GAP +
    maxCards * (CARD_HEIGHT + ROW_GAP) +
    COL_FOOTER_H +
    PAGE_MARGIN;

  return { canvasWidth, canvasHeight, columns };
}

// ─── Utilitários exportados ───────────────────────────────────────────────────

export function findCard(layout: LayoutData, courseCode: string): CardRect | undefined {
  for (const col of layout.columns) {
    const card = col.cards.find(c => c.courseCode === courseCode);
    if (card) return card;
  }
  return undefined;
}
